MBBSEmu/DOS/ExeRuntime.cs

```diff
@@ -0,0 +1,142 @@
+            Registers.ES = PSP_SEGMENT;
+            Registers.DS = PSP_SEGMENT;
+        }
+
+        public bool Load()
+        {
+            ApplyRelocation();
+            LoadSegments();
+            SetupPSP();
+            SetupEnvironmentVariables();
+            return true;
+        }
+
+        public void Run()
+        {
+            while (!Registers.Halt)
+                Cpu.Tick();
+        }
+
+        /// <summary>
+        ///     Applies Relocation Records to the EXE Code Segments
+        /// </summary>
+        private void ApplyRelocation()
+        {
+            //Get the Code Segment in the MZFile
+            var codeSegment = File.Segments.First(x => x.Flag == (ushort)EnumSegmentFlags.Code).Data;
+
+            //For the time being, only handle the 1st relo for the data segment
+            foreach (var relo in File.RelocationRecords)
+            {
+                //Data Segment is always the last one
+                Array.Copy(BitConverter.GetBytes((ushort)File.Segments.Count), 0, codeSegment, relo.Offset,
+                    sizeof(ushort));
+            }
+        }
+
+        /// <summary>
+        ///     Loads EXE Code Segments into the Memory Core
+        /// </summary>
+        private void LoadSegments()
+        {
+            for (ushort i = 1; i <= File.Segments.Count; i++)
+            {
+                var seg = File.Segments[i - 1];
+                switch (seg.Flag)
+                {
+                    case (ushort)EnumSegmentFlags.Code:
+                        {
+                            Memory.AddSegment(seg);
+                            break;
+                        }
+                    case (ushort)EnumSegmentFlags.Data:
+                        {
+                            Memory.AddSegment(i);
+                            Memory.SetArray(i, 0, seg.Data);
+                            break;
+                        }
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Sets up PSP for the program Execution
+        /// </summary>
+        private void SetupPSP()
+        {
+            var psp = new PSPStruct { NextSegOffset = 0x9FFF, EnvSeg = ENVIRONMENT_SEGMENT };
+            Memory.AddSegment(PSP_SEGMENT);
+            Memory.SetArray(PSP_SEGMENT, 0, psp.Data);
+
+            Memory.AllocateVariable("Int21h-PSP", sizeof(ushort));
+            Memory.SetWord("Int21h-PSP", PSP_SEGMENT);
+        }
+
+        /// <summary>
+        ///     Copies any Environment Variables to the Environment Variables Segment
+        /// </summary>
+        private void SetupEnvironmentVariables()
+        {
+            Memory.AddSegment(ENVIRONMENT_SEGMENT);
+
+            if (_environmentVariables == null)
+                return;
+
+            ushort bytesWritten = 0;
+            foreach (var v in _environmentVariables)
+            {
+                Memory.SetArray(ENVIRONMENT_SEGMENT, bytesWritten, Encoding.ASCII.GetBytes(v));
```

Comment:
does this add the trailing null? (i think I see now, probably get the null from the initial ZeroMemory allocation)

I believe this structure is doubly null terminated as well

someValue=something\0
another=whatever\0
\0\0

Reply:
The line below I increment length +1 since the array is initialized with `0x0` values. Since it's writing to a dedicated sector, it'll have trailing nulls after the final result so the double-null shouldn't be an issue.

Reply:
Works for me